src/ripper/ripcddialog.cpp

```diff
@@ -241,11 +266,57 @@ void RipCDDialog::InvertSelection() {
   }
 }
 
-void RipCDDialog::Finished() { SetWorking(false); }
+void RipCDDialog::DeviceSelected(int device_index) {
+  // disconnecting from previous loader and device, if any
+  if (loader_) disconnect(loader_, nullptr, this, nullptr);
+  if (cdda_device_) disconnect(cdda_device_.get(), nullptr, this, nullptr);
+
+  ResetDialog();
+  if (device_index < 0)
+    return;  // Invalid selection, probably no devices around
+
+  Q_ASSERT(device_index < cdda_devices_.size());
+
+  DeviceInfo* device_info = cdda_devices_[device_index];
+  std::shared_ptr<ConnectedDevice> device =
+      device_manager_.Connect(device_info);
+  cdda_device_ = std::dynamic_pointer_cast<CddaDevice>(device);
```

Comment:
Can we avoid this dynamic cast?

Reply:
I don't really see how. At least not without modifying `DeviceManager` and adding a method that explicitely returns `CddaDevice`s - and that would probably just push the `dynamic_cast` into the `DeviceManager` unless we really have a separate store of the `CddaDevice`s there internally..